crates/oxc_linter/src/rules/react/react_in_jsx_scope.rs

```diff
@@ -0,0 +1,99 @@
+use oxc_ast::AstKind;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-react(react-in-jsx-scope): 'React' must be in scope when using JSX")]
+#[diagnostic(severity(warning), help("When using JSX, `<a />` expands to `React.createElement(\"a\")`. Therefore the `React` variable must be in scope."))]
+struct ReactInJsxScopeDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct ReactInJsxScope;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow missing React when using JSX
+    ///
+    /// ### Why is this bad?
+    ///
+    /// When using JSX, `<a />` expands to `React.createElement("a")`. Therefore the `React` variable must be in scope.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// var a = <a />;
+    ///
+    /// // Good
+    /// import React from "react";
+    /// var a = <a />;
+    ///
+    /// ```
+    ReactInJsxScope,
+    suspicious
+);
+
+impl Rule for ReactInJsxScope {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let node_span = match node.kind() {
+            AstKind::JSXOpeningElement(v) => v.span,
+            AstKind::JSXFragment(v) => v.span,
+            _ => return,
+        };
+
+        let scope = ctx.scopes();
+
+        if !scope
+            .ancestors(node.scope_id())
+            .any(|v| scope.get_bindings(v).iter().any(|(k, _)| k.as_str() == "React"))
```

Comment:
It feels like there is a O(1) method for this: 

On the top scope, check and see if there is a "React" symbol.

Reply:
don't think that would work if you had required react in an inner scope?

```js
import _ from 'lodash'

const test = () => {
    const React = require('react');
    
    return <div></div>
}
```

is there a way to get all symbols in the current scope?

Reply:
You can get all the bindings via `get_bindings` on the scope. 

I think we can have your implementation on the slow path, and the O(1) check on the fast path.

Reply:
So we should add early bail-out if we have "React" symbol in root scope?

Reply:
yes